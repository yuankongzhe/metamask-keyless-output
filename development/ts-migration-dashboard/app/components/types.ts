export type BoxRect = {
  moduleId: string;
  top: number;
  left: number;
  width: number;
  height: number;
  centerX: number;
  centerY: number;
};

export type BoxModel = {
  rect: BoxRect;
  dependencyBoxRects: BoxRect[];
  dependentBoxRects: BoxRect[];
};
