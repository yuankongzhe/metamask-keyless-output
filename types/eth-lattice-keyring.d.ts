declare module 'eth-lattice-keyring' {
  export default class LatticeKeyring {
    static type: string;

    appName: string | undefined;

    constructor(opts);

    _getCreds(): Promise<{
      deviceID: string;
      password: string;
      endpoint: string;
    }>;
  }
}
