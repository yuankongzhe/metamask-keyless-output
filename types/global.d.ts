// In order for variables to be considered on the global scope they must be
// declared using var and not const or let, which is why this rule is disabled
/* eslint-disable no-var */
import * as Sentry from '@sentry/browser';
import {
  Success,
  Unsuccessful,
  PROTO,
  EthereumSignedTx,
  Params,
  EthereumSignTransaction,
  EthereumSignTypedHash,
  EthereumSignMessage,
  EthereumSignTypedDataTypes,
} from '@trezor/connect-web';
import {
  OffscreenCommunicationTarget,
  TrezorAction,
} from 'shared/constants/offscreen-communication';

declare class Platform {
  openTab: (opts: { url: string }) => void;

  closeCurrentWindow: () => void;
}

declare class MessageSender {
  documentId?: string;

  documentLivecycle?: string;

  frameId?: number;

  id?: string;

  origin?: string;

  url?: string;
}

export interface LedgerIframeMissingResponse {
  success: false;
  payload: {
    error: Error;
  };
}

type ResponseType =
  | Unsuccessful
  | Success<{ publicKey: string; chainCode: string }>
  | Success<EthereumSignedTx>
  | Success<PROTO.MessageSignature>
  | Success<PROTO.EthereumTypedDataSignature>
  | Record<string, unknown>
  | LedgerIframeMissingResponse;

/**
 * Defines an overloaded set of function call signatures for the chrome
 * runtime sendMessage function. Each of these are overloaded by specific
 * input values so that the correct type can be inferred in the callback
 * method
 */
interface sendMessage {
  (
    extensionId: string,
    message: Record<string, unknown>,
    options?: Record<string, unknown>,
    callback?: (response: Record<string, unknown>) => void,
  ): void;
  (
    message: any,
    options?: Record<string, unknown>,
    callback?: (response: Record<string, unknown>) => void,
  ): void;
  <T extends EthereumSignTypedDataTypes>(
    message: {
      target: OffscreenCommunicationTarget.trezorOffscreen;
      action: TrezorAction.signTypedData;
      params: Params<EthereumSignTypedHash<T>>;
    },
    callback: (
      response: Unsuccessful | Success<PROTO.EthereumTypedDataSignature>,
    ) => void,
  ): Promise<Unsuccessful | Success<PROTO.EthereumTypedDataSignature>>;
  (
    message: {
      target: OffscreenCommunicationTarget.trezorOffscreen;
      action: TrezorAction.signTransaction;
      params: Params<EthereumSignTransaction>;
    },
    callback: (response: Unsuccessful | Success<EthereumSignedTx>) => void,
  ): Promise<Unsuccessful | Success<EthereumSignedTx>>;
  (
    message: {
      target: OffscreenCommunicationTarget.trezorOffscreen;
      action: TrezorAction.signMessage;
      params: Params<EthereumSignMessage>;
    },
    callback: (
      response: Unsuccessful | Success<PROTO.MessageSignature>,
    ) => void,
  ): Promise<Unsuccessful | Success<PROTO.MessageSignature>>;
  (
    message: {
      target: OffscreenCommunicationTarget.trezorOffscreen;
      action: TrezorAction.getPublicKey;
      params: { path: string; coin: string };
    },
    callback: (
      response:
        | Unsuccessful
        | Success<{ publicKey: string; chainCode: string }>,
    ) => void,
  ): Promise<Unsuccessful | Success<{ publicKey: string; chainCode: string }>>;
  (
    message: {
      target: OffscreenCommunicationTarget.ledgerOffscreen;
      action: LedgerAction.signTransaction;
      params: { hdPath: string; tx: string };
    },
    callback: (response: {
      success: boolean;
      payload: { v: string; s: string; r: string; error?: Error };
    }) => void,
  ): Promise<{
    success: boolean;
    payload?: { v: string; s: string; r: string };
  }>;
  (
    message:
      | {
          target: OffscreenCommunicationTarget.ledgerOffscreen;
          action: LedgerAction.signMessage;
          params: { hdPath: string; message: string };
        }
      | {
          target: OffscreenCommunicationTarget.ledgerOffscreen;
          action: LedgerAction.signTypedData;
          params: {
            hdPath: string;
            domainSeparatorHex: string;
            hashStructMessageHex: string;
          };
        },
    callback: (response: {
      success: boolean;
      payload: {
        v: number;
        s: string;
        r: string;
        error?: Error;
      };
    }) => void,
  ): Promise<{ v: number; s: string; r: string }>;
  (
    message: {
      target: OffscreenCommunicationTarget.ledgerOffscreen;
      action: LedgerAction.getPublicKey;
      params: { hdPath: string };
    },
    callback: (response: {
      success: boolean;
      payload: {
        publicKey: string;
        address: string;
        chainCode?: string;
        error?: Error;
      };
    }) => void,
  ): Promise<{ publicKey: string; address: string; chainCode?: string }>;
  (
    message: {
      target: OffscreenCommunicationTarget.ledgerOffscreen;
      action: LedgerAction.updateTransport;
      params: { transportType: string };
    },
    callback: (response: { success: boolean }) => void,
  ): Promise<boolean>;
  (
    message: {
      target: OffscreenCommunicationTarget.ledgerOffscreen;
      action: LedgerAction.makeApp;
    },
    callback: (response: { success: boolean; error?: Error }) => void,
  ): Promise<boolean>;
  (
    message: {
      target: OffscreenCommunicationTarget.latticeOffscreen;
      params: {
        url: string;
      };
    },
    callback: (response: { result: any; error?: Error }) => void,
  );
  (
    message: Record<string, unknown>,
    callback?: (response: ResponseType) => void,
  ): void;
}

declare class Runtime {
  onMessage: {
    addListener: (
      callback: (
        message: any,
        sender: MessageSender,
        sendResponse: (response?: ResponseType) => void,
      ) => void,
    ) => void;
  };

  sendMessage: sendMessage;
}

declare class Chrome {
  runtime: Runtime;
}

type SentryObject = Sentry & {
  // Verifies that the user has opted into metrics and then updates the sentry
  // instance to track sessions and begins the session.
  startSession: () => void;

  // Verifies that the user has opted out of metrics and then updates the
  // sentry instance to NOT track sessions and ends the current session.
  endSession: () => void;

  // Calls either startSession or endSession based on optin status
  toggleSession: () => void;
};

export declare global {
  var platform: Platform;
  // Sentry is undefined in dev, so use optional chaining
  var sentry: SentryObject | undefined;

  var chrome: Chrome;

  namespace jest {
    interface Matchers<R> {
      toBeFulfilled(): Promise<R>;
      toNeverResolve(): Promise<R>;
    }
  }
}
