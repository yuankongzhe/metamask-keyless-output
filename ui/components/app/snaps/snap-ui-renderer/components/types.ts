import { Component } from '@metamask/snaps-sdk';

export type UIComponentParams<T extends Component> = {
  map: Record<string, number>;
  element: T;
  form?: string;
};

export type UIComponent = {
  element: string;
  props?: Record<string, unknown>;
  children?: UIComponent[] | string;
  key?: string;
};

export type UIComponentFactory<T extends Component> = (
  params: UIComponentParams<T>,
) => UIComponent;
