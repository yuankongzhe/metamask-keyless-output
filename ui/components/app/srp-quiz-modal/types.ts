import { ButtonSize, ButtonVariant } from '../../component-library';

export enum QuizStage {
  introduction = 'introduction',
  questionOne = 'question_one',
  wrongAnswerQuestionOne = 'wrong_answer_question_one',
  rightAnswerQuestionOne = 'right_answer_question_one',
  questionTwo = 'question_two',
  wrongAnswerQuestionTwo = 'wrong_answer_question_two',
  rightAnswerQuestionTwo = 'right_answer_question_two',
}

export interface IQuizInformationProps {
  /**
   * The icon to display in the modal should use <Icon /> component
   */
  icon?: any;
  /**
   * The image to display in the modal
   */
  image?: string;
  /**
   * The text content to go inside of the <Text /> component
   */
  content: string;
  /**
   * More text content to go inside of the <Text /> component
   */
  moreContent?: string;
  /**
   * Array of <Button /> component props
   */
  buttons: {
    onClick: () => void;
    label: string;
    variant: ButtonVariant;
    size?: ButtonSize;
    'data-testid'?: string;
  }[];
}

export type JSXDict = { [key: string]: () => JSX.Element };
