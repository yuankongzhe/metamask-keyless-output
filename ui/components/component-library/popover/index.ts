export { Popover } from './popover';
export { PopoverPosition, PopoverRole } from './popover.types';
export type {
  PopoverProps,
  PopoverStyleUtilityProps,
  PopoverComponent,
} from './popover.types';
