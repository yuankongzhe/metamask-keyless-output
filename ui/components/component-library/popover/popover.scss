
.mm-popover {
  box-shadow: var(--shadow-size-md) var(--color-shadow-default);

  /* Hide the popper when the reference is hidden */
  &--reference-hidden[data-popper-reference-hidden="true"] {
    visibility: hidden;
    pointer-events: none;

    > .mm-popover__arrow::before {
      visibility: hidden;
    }
  }
}

.mm-popover__arrow,
.mm-popover__arrow::before {
  position: absolute;
  width: 8px;
  height: 8px;
  background: inherit;
}

.mm-popover__arrow {
  width: 40px;
  height: 40px;
  visibility: hidden;
}

.mm-popover__arrow::before {
  display: block;
  background-color: inherit;
  border: 1px solid;
  border-left-color: inherit;
  border-top-color: inherit;
  border-bottom-color: transparent;
  border-right-color: transparent;
  visibility: visible;
  content: '';
  transform: rotate(45deg);
  border-radius: 2px  0 0 0;
}

.mm-popover[data-popper-placement^='top'] > .mm-popover__arrow {
  bottom: -20px;

  &::before {
    transform: rotate(-135deg);
  }
}

.mm-popover[data-popper-placement^='bottom'] > .mm-popover__arrow {
  top: -20px;
}

.mm-popover[data-popper-placement^='left'] > .mm-popover__arrow {
  right: -20px;

  &::before {
    transform: rotate(135deg);
  }
}

.mm-popover[data-popper-placement^='right'] > .mm-popover__arrow {
  left: -20px;

  &::before {
    transform: rotate(-45deg);
  }
}
