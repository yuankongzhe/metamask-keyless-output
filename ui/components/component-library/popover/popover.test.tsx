/* eslint-disable jest/require-top-level-describe */
import { render, fireEvent } from '@testing-library/react';
import React, { useState } from 'react';
import { Popover } from './popover';
import { PopoverPosition } from './popover.types';

describe('Popover', () => {
  it('should render popover element correctly', () => {
    const { getByTestId, getByText, container } = render(
      <Popover data-testid="popover" isOpen={true} isPortal={false}>
        Popover
      </Popover>,
    );
    expect(getByText('Popover')).toBeDefined();
    expect(container.querySelector('popover')).toBeDefined();
    expect(getByTestId('popover')).toHaveClass('mm-popover');
    expect(container).toMatchSnapshot();
  });

  it('should NOT render popover', () => {
    const { queryByTestId } = render(
      <>
        <Popover isOpen={false}>Popover not open</Popover>
      </>,
    );
    expect(queryByTestId('popover')).not.toBeInTheDocument();
  });

  it('should render popover children', () => {
    const { getByText } = render(
      <Popover isOpen={true}>Popover content goes here</Popover>,
    );
    expect(getByText('Popover content goes here')).toBeDefined();
  });

  it('should render popover position', () => {
    const { getByTestId } = render(
      <>
        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.Auto}
          position={PopoverPosition.Auto}
        >
          {PopoverPosition.Auto}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid="top"
          position={PopoverPosition.Top}
        >
          {PopoverPosition.Top}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.TopStart}
          position={PopoverPosition.TopStart}
        >
          {PopoverPosition.TopStart}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.TopEnd}
          position={PopoverPosition.TopEnd}
        >
          {PopoverPosition.TopEnd}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.Right}
          position={PopoverPosition.Right}
        >
          {PopoverPosition.Right}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.RightStart}
          position={PopoverPosition.RightStart}
        >
          {PopoverPosition.RightStart}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.RightEnd}
          position={PopoverPosition.RightEnd}
        >
          {PopoverPosition.RightEnd}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.Bottom}
          position={PopoverPosition.Bottom}
        >
          {PopoverPosition.Bottom}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.BottomStart}
          position={PopoverPosition.BottomStart}
        >
          {PopoverPosition.BottomStart}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.BottomEnd}
          position={PopoverPosition.BottomEnd}
        >
          {PopoverPosition.BottomEnd}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.Left}
          position={PopoverPosition.Left}
        >
          {PopoverPosition.Left}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.LeftStart}
          position={PopoverPosition.LeftStart}
        >
          {PopoverPosition.LeftStart}
        </Popover>

        <Popover
          isOpen={true}
          isPortal={false}
          data-testid={PopoverPosition.LeftEnd}
          position={PopoverPosition.LeftEnd}
        >
          {PopoverPosition.LeftEnd}
        </Popover>
      </>,
    );

    expect(getByTestId(PopoverPosition.Auto)).toBeDefined();
    expect(getByTestId(PopoverPosition.Top)).toBeDefined();
    expect(getByTestId(PopoverPosition.TopStart)).toBeDefined();
    expect(getByTestId(PopoverPosition.TopEnd)).toBeDefined();
    expect(getByTestId(PopoverPosition.Right)).toBeDefined();
    expect(getByTestId(PopoverPosition.RightStart)).toBeDefined();
    expect(getByTestId(PopoverPosition.RightEnd)).toBeDefined();
    expect(getByTestId(PopoverPosition.Bottom)).toBeDefined();
    expect(getByTestId(PopoverPosition.BottomStart)).toBeDefined();
    expect(getByTestId(PopoverPosition.BottomEnd)).toBeDefined();
    expect(getByTestId(PopoverPosition.Left)).toBeDefined();
    expect(getByTestId(PopoverPosition.LeftStart)).toBeDefined();
    expect(getByTestId(PopoverPosition.LeftEnd)).toBeDefined();
  });

  it('should render an arrow on popover', () => {
    const { getByTestId } = render(
      <Popover data-testid="popover" isOpen={true} hasArrow>
        Popover
      </Popover>,
    );

    const arrowElement =
      getByTestId('popover').querySelector('.mm-popover__arrow');
    expect(arrowElement).toHaveClass('mm-popover__arrow');
  });

  test('should render Popover with isPortal set to false', () => {
    const { getByTestId } = render(
      <div>
        <Popover
          data-testid="popover"
          isOpen={true}
          position={PopoverPosition.Bottom}
          isPortal={false}
        >
          <p>Popover content</p>
        </Popover>
      </div>,
    );
    // Check that the Popover is rendered inside the body DOM
    expect(getByTestId('popover')).toBeInTheDocument();
  });

  test('should render Popover with isPortal set to true', () => {
    const { getByTestId } = render(
      <div>
        <Popover data-testid="popover" isOpen={true} isPortal={true}>
          <p>Popover content</p>
        </Popover>
      </div>,
    );

    expect(getByTestId('popover')).toBeTruthy();
  });

  test('should add reference-hidden classname when referenceHidden prop is true', () => {
    const { getByTestId } = render(
      <div>
        <Popover data-testid="popover" isOpen={true} referenceHidden={true}>
          <p>Popover content</p>
        </Popover>
      </div>,
    );

    expect(getByTestId('popover')).toHaveClass('mm-popover--reference-hidden');
  });

  const EscKeyTestComponent = () => {
    const [isOpen, setIsOpen] = useState(true);

    return (
      <Popover
        isOpen={isOpen}
        referenceHidden={false}
        onPressEscKey={() => setIsOpen(false)}
      >
        Press esc key to close
      </Popover>
    );
  };

  test('Press esc key to close fires', () => {
    // Render the component
    const { getByText, queryByText } = render(<EscKeyTestComponent />);

    // Assert that the popover is initially visible
    expect(getByText('Press esc key to close')).toBeVisible();

    // Trigger the "Escape" key press event
    fireEvent.keyDown(document, { key: 'Escape' });

    // Assert that the popover closes
    expect(queryByText('Press esc key to close')).not.toBeInTheDocument();
  });

  const ClickOutsideTestComponent = () => {
    const [isOpen, setIsOpen] = useState(true);

    const handleOnClickOutside = () => {
      setIsOpen(false);
    };

    return (
      <div>
        <Popover
          isOpen={isOpen}
          referenceHidden={false}
          onClickOutside={handleOnClickOutside}
        >
          Click outside to close
        </Popover>
        <div data-testid="outside-click-target">Click outside</div>
      </div>
    );
  };

  test('Close popover when clicking outside using onClickOutside prop', () => {
    const { getByText, getByTestId, queryByText } = render(
      <ClickOutsideTestComponent />,
    );

    // Assert that the popover is initially open
    expect(getByText('Click outside to close')).toBeVisible();

    // Simulate a click event outside the popover
    const outsideClickTarget = getByTestId('outside-click-target');
    fireEvent.click(outsideClickTarget);

    // Assert that the popover is closed after the click event
    expect(queryByText('Click outside to close')).not.toBeInTheDocument();
  });
});
