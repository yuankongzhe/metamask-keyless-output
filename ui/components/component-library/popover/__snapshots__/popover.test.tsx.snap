// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render popover element correctly 1`] = `
<div>
  <div
    class="mm-box mm-popover mm-popover--open mm-popover--reference-hidden mm-box--padding-4 mm-box--background-color-background-default mm-box--rounded-lg mm-box--border-color-border-muted box--border-style-solid box--border-width-1"
    data-testid="popover"
    role="tooltip"
    style="position: absolute; left: 0px; top: 0px; width: auto;"
  >
    Popover
  </div>
</div>
`;
